Ignore unknown flags in clang-tidy

Now, `cmake` has started passing gcc-specific flags for modules support, which breaks `clang-tidy`. Is there a way to tell clang-tidy to ignore the flags it doesn't understand instead of erroring out?

https://gitlab.kitware.com/cmake/cmake/-/issues/26283

```shell
FAILED: my_exe/test/CMakeFiles/my_exe_lib_tests.dir/Debug/tests.cpp.o 
/opt/homebrew/Cellar/cmake/3.30.3/bin/cmake -E __run_co_compile --launcher=/opt/homebrew/bin/ccache --tidy="/Users/runner/llvm/bin/clang-tidy;-extra-arg=-Wno-unknown-warning-option;--extra-arg-before=--driver-mode=g++" --source=/Users/runner/work/cpp_vcpkg_project/cpp_vcpkg_project/my_exe/test/tests.cpp -- /opt/homebrew/bin/g++-14 -DDO_NOT_USE_WMAIN -DCMAKE_INTDIR=\"Debug\" -I/Users/runner/work/cpp_vcpkg_project/cpp_vcpkg_project/my_exe/include -isystem /Users/runner/work/cpp_vcpkg_project/cpp_vcpkg_project/build/vcpkg_installed/arm64-osx/include -g -std=c++20 -arch arm64 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -fdiagnostics-color=always -Wall -Wextra -Wextra-semi -Wshadow -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion -Wnull-dereference -Wdouble-promotion -Wformat=2 -Wimplicit-fallthrough -Wmisleading-indentation -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wuseless-cast --coverage -O0 -g -MD -MT my_exe/test/CMakeFiles/my_exe_lib_tests.dir/Debug/tests.cpp.o -MF my_exe/test/CMakeFiles/my_exe_lib_tests.dir/Debug/tests.cpp.o.d -fmodules-ts -fmodule-mapper=my_exe/test/CMakeFiles/my_exe_lib_tests.dir/Debug/tests.cpp.o.modmap -MD -fdeps-format=p1689r5 -x c++ -o my_exe/test/CMakeFiles/my_exe_lib_tests.dir/Debug/tests.cpp.o -c /Users/runner/work/cpp_vcpkg_project/cpp_vcpkg_project/my_exe/test/tests.cpp
error: unknown argument: '-fdeps-format=p1689r5' [clang-diagnostic-error]
error: unknown argument: '-fmodule-mapper=my_exe/test/CMakeFiles/my_exe_lib_tests.dir/Debug/tests.cpp.o.modmap' [clang-diagnostic-error]
error: unknown argument: '-fmodules-ts' [clang-diagnostic-error]

```

This is also related to https://github.com/clangd/clangd/issues/662
Perhaps a `remove-flag` option could solve the problem.

--- a/clang-tools-extra/clang-tidy/ClangTidy.cpp
+++ b/clang-tools-extra/clang-tidy/ClangTidy.cpp
@@ -586,6 +586,24 @@ runClangTidy(clang::tidy::ClangTidyContext &Context,
         return AdjustedArgs;
       };
 
+  // Remove unwanted arguments passed to the compiler
+  const ArgumentsAdjuster PerFileArgumentRemover =
+      [&Context](const CommandLineArguments &Args, StringRef Filename) {
+        ClangTidyOptions Opts = Context.getOptionsForFile(Filename);
+        CommandLineArguments AdjustedArgs = Args;
+
+        if (Opts.RemovedArgs) {
+          for (const StringRef ArgToRemove : *Opts.RemovedArgs) {
+            AdjustedArgs.erase(std::remove(AdjustedArgs.begin(),
+                                           AdjustedArgs.end(), ArgToRemove),
+                               AdjustedArgs.end());
+          }
+        }
+
+        return AdjustedArgs;
+      };
+
+  Tool.appendArgumentsAdjuster(PerFileArgumentRemover);
   Tool.appendArgumentsAdjuster(PerFileExtraArgumentsInserter);
   Tool.appendArgumentsAdjuster(getStripPluginsAdjuster());
   Context.setEnableProfiling(EnableCheckProfile);

--- a/clang-tools-extra/clang-tidy/ClangTidyOptions.cpp
+++ b/clang-tools-extra/clang-tidy/ClangTidyOptions.cpp
@@ -232,6 +232,7 @@ template <> struct MappingTraits<ClangTidyOptions> {
     IO.mapOptional("CheckOptions", Options.CheckOptions);
     IO.mapOptional("ExtraArgs", Options.ExtraArgs);
     IO.mapOptional("ExtraArgsBefore", Options.ExtraArgsBefore);
+    IO.mapOptional("RemovedArgs", Options.RemovedArgs);
     IO.mapOptional("InheritParentConfig", Options.InheritParentConfig);
     IO.mapOptional("UseColor", Options.UseColor);
     IO.mapOptional("SystemHeaders", Options.SystemHeaders);
@@ -254,6 +255,7 @@ ClangTidyOptions ClangTidyOptions::getDefaults() {
   Options.SystemHeaders = false;
   Options.FormatStyle = "none";
   Options.User = std::nullopt;
+  Options.RemovedArgs = std::nullopt;
   for (const ClangTidyModuleRegistry::entry &Module :
        ClangTidyModuleRegistry::entries())
     Options.mergeWith(Module.instantiate()->getModuleOptions(), 0);
@@ -297,6 +299,7 @@ ClangTidyOptions &ClangTidyOptions::mergeWith(const ClangTidyOptions &Other,
   overrideValue(UseColor, Other.UseColor);
   mergeVectors(ExtraArgs, Other.ExtraArgs);
   mergeVectors(ExtraArgsBefore, Other.ExtraArgsBefore);
+  mergeVectors(RemovedArgs, Other.RemovedArgs);
   // FIXME: how to handle duplicate names check?
   mergeVectors(CustomChecks, Other.CustomChecks);
   for (const auto &KeyValue : Other.CheckOptions) {

--- a/clang-tools-extra/clang-tidy/ClangTidyOptions.h
+++ b/clang-tools-extra/clang-tidy/ClangTidyOptions.h
@@ -85,7 +85,7 @@ struct ClangTidyOptions {
   /// main files will always be displayed.
   std::optional<std::string> HeaderFilterRegex;
 
-  /// \brief Exclude warnings from headers matching this filter, even if they
+  /// Exclude warnings from headers matching this filter, even if they
   /// match \c HeaderFilterRegex.
   std::optional<std::string> ExcludeHeaderFilterRegex;
 
@@ -150,6 +150,9 @@ struct ClangTidyOptions {
 
   /// Add extra compilation arguments to the start of the list.
   std::optional<ArgList> ExtraArgsBefore;
+
+  /// Remove command line arguments sent to the compiler matching this.
+  std::optional<ArgList> RemovedArgs;
 
   /// Only used in the FileOptionsProvider and ConfigOptionsProvider. If true
   /// and using a FileOptionsProvider, it will take a configuration file in the

--- a/clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp
+++ b/clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp
@@ -362,7 +362,7 @@ static cl::list<std::string> RemovedArgs("removed-arg", desc(R"(
 List of arguments to remove from the command
 line sent to the compiler. Please note that
 removing arguments might change the semantic
-of the analzed code, possibly leading to
+of the analyzed code, possibly leading to
 compiler errors, false positives or
 false negatives. This option is applied 
 before --extra-arg and --extra-arg-before)"),

--- a/clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp
+++ b/clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp
@@ -18,6 +18,7 @@
 #include "../ClangTidy.h"
 #include "../ClangTidyForceLinker.h"
 #include "../GlobList.h"
+#include "../utils/OptionsUtils.h"
 #include "clang/Tooling/CommonOptionsParser.h"
 #include "llvm/ADT/StringSet.h"
 #include "llvm/Support/CommandLine.h"
@@ -77,6 +78,7 @@ Configuration files:
                                  (if any exists) will be taken and the current
                                  config file will be applied on top of the
                                  parent one.
+  RemovedArgs                  - Same as '--removed-arg'.
   SystemHeaders                - Same as '--system-headers'.
   UseColor                     - Same as '--use-color'.
   User                         - Specifies the name or e-mail of the user
@@ -356,6 +358,16 @@ see https://clang.llvm.org/extra/clang-tidy/QueryBasedCustomChecks.html.
                                               cl::init(false),
                                               cl::cat(ClangTidyCategory));
 
+static cl::list<std::string> RemovedArgs("removed-arg", desc(R"(
+List of arguments to remove from the command
+line sent to the compiler. Please note that
+removing arguments might change the semantic
+of the analzed code, possibly leading to
+compiler errors, false positives or
+false negatives. This option is applied 
+before --extra-arg and --extra-arg-before)"),
+                                         cl::cat(ClangTidyCategory));
+
 namespace clang::tidy {
 
 static void printStats(const ClangTidyStats &Stats) {
@@ -422,6 +434,8 @@ createOptionsProvider(llvm::IntrusiveRefCntPtr<vfs::FileSystem> FS) {
     OverrideOptions.FormatStyle = FormatStyle;
   if (UseColor.getNumOccurrences() > 0)
     OverrideOptions.UseColor = UseColor;
+  if (RemovedArgs.getNumOccurrences() > 0)
+    OverrideOptions.RemovedArgs = RemovedArgs;
 
   auto LoadConfig =
       [&](StringRef Configuration,

--- a/clang-tools-extra/clang-tidy/tool/clang-tidy-diff.py
+++ b/clang-tools-extra/clang-tidy/tool/clang-tidy-diff.py
@@ -229,6 +229,13 @@ def main():
         help="Additional argument to prepend to the compiler " "command line.",
     )
     parser.add_argument(
+        "-removed-arg",
+        dest="removed_arg",
+        action="append",
+        default=[],
+        help="Arguments to remove from the compiler command line.",
+    )
+    parser.add_argument(
         "-quiet",
         action="store_true",
         default=False,
@@ -378,6 +385,8 @@ def main():
         common_clang_tidy_args.append("-extra-arg=%s" % arg)
     for arg in args.extra_arg_before:
         common_clang_tidy_args.append("-extra-arg-before=%s" % arg)
+    for arg in args.removed_arg:
+        common_clang_tidy_args.append("-removed-arg=%s" % arg)
     for plugin in args.plugins:
         common_clang_tidy_args.append("-load=%s" % plugin)
     if args.warnings_as_errors:

--- a/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
+++ b/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
@@ -96,6 +96,7 @@ def get_tidy_invocation(
     allow_enabling_alpha_checkers: bool,
     extra_arg: List[str],
     extra_arg_before: List[str],
+    removed_arg: List[str],
     quiet: bool,
     config_file_path: str,
     config: str,
@@ -135,6 +136,8 @@ def get_tidy_invocation(
         start.append(f"-extra-arg={arg}")
     for arg in extra_arg_before:
         start.append(f"-extra-arg-before={arg}")
+    for arg in removed_arg:
+        start.append(f"-removed-arg={arg}")
     start.append(f"-p={build_path}")
     if quiet:
         start.append("-quiet")
@@ -377,6 +380,7 @@ async def run_tidy(
         args.allow_enabling_alpha_checkers,
         args.extra_arg,
         args.extra_arg_before,
+        args.removed_arg,
         args.quiet,
         args.config_file,
         args.config,
@@ -552,6 +556,13 @@ async def main() -> None:
         help="Additional argument to prepend to the compiler command line.",
     )
     parser.add_argument(
+        "-removed-arg",
+        dest="removed_arg",
+        action="append",
+        default=[],
+        help="Arguments to remove from the compiler command line.",
+    )
+    parser.add_argument(
         "-quiet", action="store_true", help="Run clang-tidy in quiet mode."
     )
     parser.add_argument(
@@ -638,6 +649,7 @@ async def main() -> None:
             args.allow_enabling_alpha_checkers,
             args.extra_arg,
             args.extra_arg_before,
+            args.removed_arg,
             args.quiet,
             args.config_file,
             args.config,

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -192,6 +192,11 @@ Improvements to clang-tidy
   From now on, users should specify explicitly that they want CSA checks to run
   in :program:`clang-tidy`.
 
+- Improved :program:`clang-tidy` by adding the `--removed-arg` option to remove
+  arguments sent to the compiler when invoking Clang-Tidy. This option was also
+  added to :program:`run-clang-tidy.py` and :program:`clang-tidy-diff.py` and 
+  can be configured in the config file through the `RemovedArgs` option.
+
 - Deprecated the :program:`clang-tidy` ``zircon`` module. All checks have been
   moved to the ``fuchsia`` module instead. The ``zircon`` module will be removed
   in the 24th release.

--- a/clang-tools-extra/docs/clang-tidy/index.rst
+++ b/clang-tools-extra/docs/clang-tidy/index.rst
@@ -331,6 +331,7 @@ An overview of all the command-line options:
                                    example, to place the correct user name in
                                    TODO() comments in the relevant check.
     WarningsAsErrors             - Same as '--warnings-as-errors'.
+    RemovedArgs                  - Same as '--removed-arg'
 
     The effective configuration can be inspected using --dump-config:
 

--- a/clang-tools-extra/test/clang-tidy/infrastructure/invalid-command-line.cpp
+++ b/clang-tools-extra/test/clang-tidy/infrastructure/invalid-command-line.cpp
@@ -1,4 +1,4 @@
 // RUN: not clang-tidy --invalid-arg 2>&1 | FileCheck %s
 
 // CHECK: error: clang-tidy{{(\.exe)?}}: Unknown command line argument '--invalid-arg'.  Try: '{{.*}}clang-tidy{{(\.exe)?}} --help'
-// CHECK-NEXT: clang-tidy{{(\.exe)?}}: Did you mean '--extra-arg'?
+// CHECK-NEXT: clang-tidy{{(\.exe)?}}: Did you mean '--removed-arg'?

--- a/clang-tools-extra/test/clang-tidy/infrastructure/removed-args.cpp
+++ b/clang-tools-extra/test/clang-tidy/infrastructure/removed-args.cpp
@@ -1,8 +1,8 @@
-// RUN: not clang-tidy %s -- -fnot-an-option | FileCheck %s -check-prefix=INVALID-A
-// RUN: clang-tidy %s --config="{RemovedArgs: ['-fnot-an-option']}" -- -fnot-an-option
-// RUN: clang-tidy %s --config="{RemovedArgs: ['-fnot-another-option', '-fnot-an-option']}" -- -fnot-an-option -fnot-another-option
-// RUN: clang-tidy %s --removed-arg="-fnot-an-option" -- -fnot-an-option
-// RUN: not clang-tidy %s --removed-arg="-fnot-an-option" -- -fnot-an-option -fnot-another-option | FileCheck %s -check-prefix=INVALID-B
+// RUN: not clang-tidy %s --allow-no-checks -- -fnot-an-option | FileCheck %s -check-prefix=INVALID-A
+// RUN: clang-tidy %s --allow-no-checks --config="{RemovedArgs: ['-fnot-an-option']}" -- -fnot-an-option
+// RUN: clang-tidy %s --allow-no-checks --config="{RemovedArgs: ['-fnot-another-option', '-fnot-an-option']}" -- -fnot-an-option -fnot-another-option
+// RUN: clang-tidy %s --allow-no-checks --removed-arg="-fnot-an-option" -- -fnot-an-option
+// RUN: not clang-tidy %s --allow-no-checks --removed-arg="-fnot-an-option" -- -fnot-an-option -fnot-another-option | FileCheck %s -check-prefix=INVALID-B
 
 // INVALID-A: error: unknown argument: '-fnot-an-option' [clang-diagnostic-error]
 // INVALID-B: error: unknown argument: '-fnot-another-option' [clang-diagnostic-error]

--- a/clang-tools-extra/test/clang-tidy/infrastructure/removed-args.cpp
+++ b/clang-tools-extra/test/clang-tidy/infrastructure/removed-args.cpp
@@ -1,8 +1,8 @@
-// RUN: not clang-tidy %s --allow-no-checks -- -fnot-an-option | FileCheck %s -check-prefix=INVALID-A
-// RUN: clang-tidy %s --allow-no-checks --config="{RemovedArgs: ['-fnot-an-option']}" -- -fnot-an-option
-// RUN: clang-tidy %s --allow-no-checks --config="{RemovedArgs: ['-fnot-another-option', '-fnot-an-option']}" -- -fnot-an-option -fnot-another-option
-// RUN: clang-tidy %s --allow-no-checks --removed-arg="-fnot-an-option" -- -fnot-an-option
-// RUN: not clang-tidy %s --allow-no-checks --removed-arg="-fnot-an-option" -- -fnot-an-option -fnot-another-option | FileCheck %s -check-prefix=INVALID-B
+// RUN: not clang-tidy %s --checks="readability-function-size" -- -fnot-an-option | FileCheck %s -check-prefix=INVALID-A
+// RUN: clang-tidy %s --checks="readability-function-size" --config="{RemovedArgs: ['-fnot-an-option']}" -- -fnot-an-option
+// RUN: clang-tidy %s --checks="readability-function-size" --config="{RemovedArgs: ['-fnot-another-option', '-fnot-an-option']}" -- -fnot-an-option -fnot-another-option
+// RUN: clang-tidy %s --checks="readability-function-size" --removed-arg="-fnot-an-option" -- -fnot-an-option
+// RUN: not clang-tidy %s --checks="readability-function-size" --removed-arg="-fnot-an-option" -- -fnot-an-option -fnot-another-option | FileCheck %s -check-prefix=INVALID-B
 
 // INVALID-A: error: unknown argument: '-fnot-an-option' [clang-diagnostic-error]
 // INVALID-B: error: unknown argument: '-fnot-another-option' [clang-diagnostic-error]

--- a/clang-tools-extra/test/clang-tidy/infrastructure/removed-args.cpp
+++ b/clang-tools-extra/test/clang-tidy/infrastructure/removed-args.cpp
@@ -1,0 +1,8 @@
+// RUN: not clang-tidy %s -- -fnot-an-option | FileCheck %s -check-prefix=INVALID-A
+// RUN: clang-tidy %s --config="{RemovedArgs: ['-fnot-an-option']}" -- -fnot-an-option
+// RUN: clang-tidy %s --config="{RemovedArgs: ['-fnot-another-option', '-fnot-an-option']}" -- -fnot-an-option -fnot-another-option
+// RUN: clang-tidy %s --removed-arg="-fnot-an-option" -- -fnot-an-option
+// RUN: not clang-tidy %s --removed-arg="-fnot-an-option" -- -fnot-an-option -fnot-another-option | FileCheck %s -check-prefix=INVALID-B
+
+// INVALID-A: error: unknown argument: '-fnot-an-option' [clang-diagnostic-error]
+// INVALID-B: error: unknown argument: '-fnot-another-option' [clang-diagnostic-error]

